File: 2015/02/rs/src/main.rs
use std::env;
use std::fs;
use std::time::{Instant};
use regex::Regex;

fn part1(dimensions: &Vec<(u32, u32, u32)>) -> u32 {
    let mut total_paper = 0;
    for (w, l, h) in dimensions {
        let wl = w * l;
        let wh = w * h;
        let hl = h * l;
        let smallest = [&wl, &wh, &hl].iter().map(|x| *x).min().unwrap();
        total_paper += 2 * (wl + wh + hl) + smallest;
    }
    total_paper
}

fn part2(dimensions: &Vec<(u32, u32, u32)>) -> u32 {
    let mut total_ribbon = 0;
    for (w, l, h) in dimensions {
        let mut sides = [&w, &l, &h];
        sides.sort();
        total_ribbon += 2 * (*sides[0] + *sides[1]) + w * l * h;
    }
    total_ribbon
}

fn solve(dimensions: &Vec<(u32, u32, u32)>) -> (u32, u32) {
    (part1(dimensions), part2(dimensions))
}

fn get_input(file_path: &String) -> Vec<(u32, u32, u32)> {
    let re = Regex::new(r"^(\d+)x(\d+)x(\d+)$").unwrap();
    let processor = |c: &regex::Captures, index| c.get(index).unwrap().as_str().parse::<u32>().unwrap();
    fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .lines()
        .map(|line| re.captures(line)
            .map(|c| (processor(&c, 1), processor(&c, 2), processor(&c, 3))).unwrap())
            .collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:?}", now.elapsed().as_secs_f32());
}

File: 2015/02/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

typedef struct
{
    int width;
    int length;
    int height;
} Dimensions;

typedef struct
{
    Dimensions *dimensions;
    int count;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

#define MIN(a, b) ((a) < (b) ? (a) : (b))
#define MIN_THREE(a, b, c) MIN(MIN(a, b), c)

int part1(Input input)
{
    int totalPaper = 0;
    while (input.count--)
    {
        int wl = input.dimensions->width * input.dimensions->length;
        int wh = input.dimensions->width * input.dimensions->height;
        int hl = input.dimensions->height * input.dimensions->length;
        int smallest = MIN_THREE(wl, wh, hl);
        totalPaper += 2 * (wl + wh + hl) + smallest;
        input.dimensions++;
    }
    return totalPaper;
}

int part2(Input input)
{
    int totalRibbon = 0;
    int height, length, width;
    int smaller1, smaller2, intermediate;
    while (input.count--)
    {
        height = input.dimensions->height;
        length = input.dimensions->length;
        width = input.dimensions->width;
        if (height > length)
        {
            smaller1 = length;
            intermediate = height;
        }
        else
        {
            smaller1 = height;
            intermediate = length;
        }
        smaller2 = intermediate > width ? width : intermediate;
        totalRibbon += 2 * (smaller1 + smaller2) + height * length * width;
        input.dimensions++;
    }
    return totalRibbon;
}

Results solve(Input dimensions)
{
    return (Results){part1(dimensions), part2(dimensions)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    Dimensions *dimensions = calloc(1024, sizeof(Dimensions));
    Dimensions *current = dimensions;
    size_t len = 0;
    char *line;
    int width, length, height;
    int count = 0;
    while (getline(&line, &len, file) != EOF)
    {
        sscanf(line, "%dx%dx%d", &width, &length, &height);
        *current++ = (Dimensions){
            width, length, height};
        count++;
    }
    fclose(file);
    return (Input){dimensions, count};
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}